src/auspex/experiment.py

```diff
@@ -640,25 +640,21 @@ def shutdown(self):
         # This includes stopping the flow of data, and must be done before terminating nodes
         self.shutdown_instruments()
 
-        # try:
-        #     while True:
-        #         if any([n.is_alive() for n in self.other_nodes]):
-        #             logger.warning("Filter pipeline appears stuck. Use keyboard interrupt to terminate.")
-        #             time.sleep(2.0)
-        #         else:
-        #             break
-        # except KeyboardInterrupt as e:
-        #     for n in self.other_nodes:
-        #         n.terminate()
-
-        time.sleep(2)
-        if any([n.is_alive() for n in self.other_nodes]):
-            logger.warning('Filter pipeline stuck, terminating.')
+        try:
+            ct_max = 5 #arbitrary waiting 10 s before giving up
+            for ct in range(ct_max):
+                if any([n.is_alive() for n in self.other_nodes]):
+                    logger.warning("Filter pipeline appears stuck. Use keyboard interrupt to terminate.")
+                    time.sleep(2.0)
+                else:
+                    break
+                for n in self.other_nodes:
+                    n.terminate()
+                raise Exception('Filter pipeline stuck!')
+        except KeyboardInterrupt as e:
```

Comment:
maybe this way we can have both keyboard interrupt and an exception after enough time?

Reply:
I think this is a reasonable solution!